### DISCUSSION TITLE
Full NAT Support

### DISCUSSION THREAD

**Discussion by infinitydon** [loxilb-io/loxilb#107] [Q&A]
Hi,

I will like to know if loxilb supports Full NAT scenarios whereby the real-servers will only see loxilb as the source and destination IP.

This can help where the loxilb and the real-servers are in separate VLANs/network.

Example with ipvs:

https://www.loadbalancer.org/blog/enabling-snat-in-lvs-xt_ipvs-and-iptables/

**Comment by nik-netlox** [comment]
Hi @infinitydon,
As far as we can understand, you are referring to one-arm load-balancer mode(https://networkgeekstuff.com/networking/basic-load-balancer-scenarios-explained/). 
If yes, then it is planned for upcoming release(later this month).

**Reply by infinitydon** [reply]
Thanks for the response, could you also let me know if other features like DSR, DSR_IPIP (tunnel mode) are been considered also?

**Reply by nik-netlox** [reply]
We have planned these features in December Release.

**Comment by UltraInstinct14** [comment]
We have implemented this with two LB modes  - full-nat and one-arm with subtle differences between them. In one-arm mode, the LB itself is placed in the same broadcast domain as client and end-points. In such a case the LB does full NAT with source-IP changed to subnet IP assigned to itself.  In full-NAT mode, there is no restriction that the LB needs to be in same broadcast domain, and it selects the source IP as the service IP when sending packets towards end-points. This is considering the fact that end-points have route to this service IP (LB)  or we can also configure GoBGP to advertise the service IP path to the end-points.

* How to configure one-arm  :
```
loxicmd create lb 20.20.20.1 --tcp=2020:8080 --endpoints=100.100.100.2:1,100.100.100.3:1,100.100.100.4:1 --mode=onearm
```
* How to configure full-nat :
```
loxicmd create lb 20.20.20.1 --tcp=2020:8080 --endpoints=31.31.31.1:1,32.32.32.1:1,33.33.33.1:1 --mode=fullnat
```
* How to configure full-nat (with BGP) :
```
loxicmd create lb 20.20.20.1 --tcp=2020:8080 --endpoints=31.31.31.1:1,32.32.32.1:1,33.33.33.1:1 --mode=fullnat --bgp
```

Example scenario and config can be found [here](https://github.com/loxilb-io/loxilb/blob/main/cicd/scenario11/config.sh)

**Comment by infinitydon** [comment]
Much thanks for the effort, will try to find time to test

**Comment by infinitydon** [comment]
I just tried the fullNAT for SCTP traffic but it is not working

```
root@ip-10-0-1-17:~# ip -br a
lo               UNKNOWN        127.0.0.1/8 ::1/128
ens5             UP             10.0.1.17/24 fe80::85:deff:fe4d:7731/64
ens6             UP             10.0.3.17/24 fe80::78:1fff:fe3b:a117/64
llb0             UNKNOWN        fe80::cc1d:67ff:feda:fff1/64
root@ip-10-0-1-17:~# loxicmd create lb 20.20.20.1 --sctp=38412:38412 --endpoints=10.0.3.10:1,10.0.3.11:1 --mode=fullnat
```

```
[API] Load balancer POST API called. url : /netlox/v1/config/loadbalancer
[API] lbRules : {{20.20.20.1 38412 sctp 0 false 2 0} [{10.0.3.10 38412 1} {10.0.3.11 38412 1}]}
nat lb-rule added - 1:dst-20.20.20.1/32,proto-132,dport-38412,-do-fullnat:eip-10.0.3.10,ep-38412,w-1,alive|eip-10.0.3.11,ep-38412,w-1,alive|
1:dst-20.20.20.1/32,proto-132,dport-38412,,do-fullnat:eip-10.0.3.10,ep-38412,w-1,alive|eip-10.0.3.11,ep-38412,w-1,alive| pc 0 bc 0
1:dst-20.20.20.1/32,proto-132,dport-38412,,do-fullnat:eip-10.0.3.10,ep-38412,w-1,alive|eip-10.0.3.11,ep-38412,w-1,alive| pc 0 bc 0
##10.0.0.59:38104 -> 10.0.1.17:22 (6):0 (Aged:0:0:0)
rdir ct4 not found 10.0.0.59:38104 -> 10.0.1.17:22 (6)
1:dst-20.20.20.1/32,proto-132,dport-38412,,do-fullnat:eip-10.0.3.10,ep-38412,w-1,alive|eip-10.0.3.11,ep-38412,w-1,alive| pc 0 bc 0
1:dst-20.20.20.1/32,proto-132,dport-38412,,do-fullnat:eip-10.0.3.10,ep-38412,w-1,alive|eip-10.0.3.11,ep-38412,w-1,alive| pc 0 bc 0
1:dst-20.20.20.1/32,proto-132,dport-38412,,do-fullnat:eip-10.0.3.10,ep-38412,w-1,alive|eip-10.0.3.11,ep-38412,w-1,alive| pc 0 bc 0
1:dst-20.20.20.1/32,proto-132,dport-38412,,do-fullnat:eip-10.0.3.10,ep-38412,w-1,alive|eip-10.0.3.11,ep-38412,w-1,alive| pc 0 bc 0
1:dst-20.20.20.1/32,proto-132,dport-38412,,do-fullnat:eip-10.0.3.10,ep-38412,w-1,alive|eip-10.0.3.11,ep-38412,w-1,alive| pc 0 bc 0
1:dst-20.20.20.1/32,proto-132,dport-38412,,do-fullnat:eip-10.0.3.10,ep-38412,w-1,alive|eip-10.0.3.11,ep-38412,w-1,alive| pc 0 bc 0
1:dst-20.20.20.1/32,proto-132,dport-38412,,do-fullnat:eip-10.0.3.10,ep-38412,w-1,alive|eip-10.0.3.11,ep-38412,w-1,alive| pc 0 bc 0
1:dst-20.20.20.1/32,proto-132,dport-38412,,do-fullnat:eip-10.0.3.10,ep-38412,w-1,alive|eip-10.0.3.11,ep-38412,w-1,alive| pc 0 bc 0
1:dst-20.20.20.1/32,proto-132,dport-38412,,do-fullnat:eip-10.0.3.10,ep-38412,w-1,alive|eip-10.0.3.11,ep-38412,w-1,alive| pc 0 bc 0
1:dst-20.20.20.1/32,proto-132,dport-38412,,do-fullnat:eip-10.0.3.10,ep-38412,w-1,alive|eip-10.0.3.11,ep-38412,w-1,alive| pc 0 bc 0
1:dst-20.20.20.1/32,proto-132,dport-38412,,do-fullnat:eip-10.0.3.10,ep-38412,w-1,alive|eip-10.0.3.11,ep-38412,w-1,alive| pc 0 bc 0
1:dst-20.20.20.1/32,proto-132,dport-38412,,do-fullnat:eip-10.0.3.10,ep-38412,w-1,alive|eip-10.0.3.11,ep-38412,w-1,alive| pc 0 bc 0
```

The sctp init message is getting to the loxiLB VM but it is not responding to the request:

```
root@ip-10-0-1-17:~# tcpdump -i any host 10.0.3.71 -s 0 -nv
tcpdump: listening on any, link-type LINUX_SLL (Linux cooked v1), capture size 262144 bytes
23:10:27.915525 IP (tos 0x2,ECT(0), ttl 64, id 0, offset 0, flags [DF], proto SCTP (132), length 68)
    10.0.3.71.9487 > 20.20.20.1.38412: sctp (1) [INIT] [init tag: 609061942] [rwnd: 106496] [OS: 2] [MIS: 2] [init TSN: 655851557]
23:14:25.355978 IP (tos 0x2,ECT(0), ttl 64, id 0, offset 0, flags [DF], proto SCTP (132), length 68)
    10.0.3.71.9487 > 20.20.20.1.38412: sctp (1) [INIT] [init tag: 3472513034] [rwnd: 106496] [OS: 2] [MIS: 2] [init TSN: 3516414202]
23:16:36.671788 IP (tos 0x2,ECT(0), ttl 64, id 0, offset 0, flags [DF], proto SCTP (132), length 68)
    10.0.3.71.9487 > 20.20.20.1.38412: sctp (1) [INIT] [init tag: 3658667719] [rwnd: 106496] [OS: 2] [MIS: 2] [init TSN: 743260408]
```

What could be wrong? Was SCTP tested?

Thanks

**Comment by nik-netlox** [comment]
Hi Christopher,
Thanks for reaching out. We will look into it and get back to you.

Thanks,
Nikhil

**Reply by infinitydon** [reply]
Full context:

- Am using AWS EC2 instances
- Am not using docker, I compiled directly using the latest master branch
- Disabled IPv6

**Reply by ghost** [reply]
Hi @infinitydon 

We have verified that sctp fullnat mode works fine. It is also added in our cicd scenario. You can check ([sctp-fullnat](https://github.com/loxilb-io/loxilb/tree/main/cicd/sctp-fullnat) config etc) it you need.  I would request you to double confirm certain things:

1. git clone method for loxilb repo has changed. It is also updated [here](https://github.com/loxilb-io/loxilbdocs/blob/main/docs/run.md) 

```
git clone --recurse-submodules https://github.com/loxilb-io/loxilb.git
```
* note the recurse-submodule part

2. Check loxilb ebpf modules are loaded as expected after running loxilb

```
ntc filter show dev ens6 ingress
```

3. Linux's rp_filter option needs to be disabled for fullnat in loxilb node

```
for i in /proc/sys/net/ipv4/conf/*/rp_filter; do echo 0 > "$i"; done
```

4. There is a limitation of updating sctp checksum using eBPF for now.  So, there is a need to disable sctp checksum check at end-points and clients. This is tricky because in our testing (with Open5GS) sometimes only disabling csum at end-points work fine. But at times we needed to disable it at client too. 

```
echo Y > /sys/module/sctp/parameters/no_checksums 
```
* We are working and checking with linux kernel community to add the needed feature.

**Reply by infinitydon** [reply]
Tested with your cicd script, it works but in non-docker setup it is still not working.

I just tested with virtualbox instead, same thing is still happening, loxilb is not responding to the SCTP INIT:

```
chris@ebenezer-loxilb:~/loxicmd$ sudo tcpdump -i any host 192.168.1.125 -s 0 -nv
[sudo] password for chris:
tcpdump: listening on any, link-type LINUX_SLL (Linux cooked v1), capture size 262144 bytes
19:03:10.868744 ARP, Ethernet (len 6), IPv4 (len 4), Request who-has 192.168.1.125 tell 192.168.1.1, length 46
19:03:15.232794 IP (tos 0x2,ECT(0), ttl 64, id 0, offset 0, flags [DF], proto SCTP (132), length 68)
    192.168.1.125.56313 > 88.88.88.1.38412: sctp (1) [INIT] [init tag: 3475740250] [rwnd: 106496] [OS: 10] [MIS: 10] [init TSN: 3998352190]
19:03:18.435601 IP (tos 0x2,ECT(0), ttl 64, id 0, offset 0, flags [DF], proto SCTP (132), length 68)
    192.168.1.125.56313 > 88.88.88.1.38412: sctp (1) [INIT] [init tag: 3475740250] [rwnd: 106496] [OS: 10] [MIS: 10] [init TSN: 3998352190]
19:03:20.483606 ARP, Ethernet (len 6), IPv4 (len 4), Request who-has 192.168.1.124 tell 192.168.1.125, length 46
19:03:20.483765 ARP, Ethernet (len 6), IPv4 (len 4), Reply 192.168.1.124 is-at 08:00:27:e7:8c:fe, length 28
19:03:24.579884 IP (tos 0x2,ECT(0), ttl 64, id 0, offset 0, flags [DF], proto SCTP (132), length 68)
    192.168.1.125.56313 > 88.88.88.1.38412: sctp (1) [INIT] [init tag: 3475740250] [rwnd: 106496] [OS: 10] [MIS: 10] [init TSN: 3998352190]
19:03:34.819545 IP (tos 0x2,ECT(0), ttl 64, id 0, offset 0, flags [DF], proto SCTP (132), length 68)
    192.168.1.125.56313 > 88.88.88.1.38412: sctp (1) [INIT] [init tag: 3475740250] [rwnd: 106496] [OS: 10] [MIS: 10] [init TSN: 3998352190]
19:03:45.060648 IP (tos 0x2,ECT(0), ttl 64, id 0, offset 0, flags [DF], proto SCTP (132), length 68)
    192.168.1.125.56313 > 88.88.88.1.38412: sctp (1) [INIT] [init tag: 3475740250] [rwnd: 106496] [OS: 10] [MIS: 10] [init TSN: 3998352190]
19:03:53.030601 ARP, Ethernet (len 6), IPv4 (len 4), Request who-has 192.168.1.125 tell 192.168.1.1, length 46
19:03:55.299564 IP (tos 0x2,ECT(0), ttl 64, id 0, offset 0, flags [DF], proto SCTP (132), length 68)
    192.168.1.125.56313 > 88.88.88.1.38412: sctp (1) [INIT] [init tag: 3475740250] [rwnd: 106496] [OS: 10] [MIS: 10] [init TSN: 3998352190]
```

loxilb node setup:

```
chris@ebenezer-loxilb:~/loxicmd$ ip -br a
lo               UNKNOWN        127.0.0.1/8 ::1/128
enp0s3           UP             192.168.1.124/24 2600:4040:2929:ef00:a00:27ff:fee7:8cfe/64 fe80::a00:27ff:fee7:8cfe/64
enp0s8           UP             192.168.56.101/24 fe80::a00:27ff:fec9:9199/64
enp0s9           UP             192.168.70.1/24 fe80::a00:27ff:fe85:d9bb/64
docker0          DOWN           172.17.0.1/16
llb0             UNKNOWN        fe80::1480:96ff:fe55:7456/64

./loxicmd create lb 88.88.88.1 --sctp=38412:38412 --endpoints=192.168.70.3:1 --mode=fullnat
```

Below is the logs

[loxilb-sctp-nat-1.txt](https://github.com/loxilb-io/loxilb/files/10149282/loxilb-sctp-nat-1.txt)

Was the setup tested using VMs directly and not docker containers?

**Reply by ghost** [reply]
We did test with VMs a while back (but not full nat mode). And  only docker based test setup is part of CICD. We will retest VM scenario. What applications are you using for sctp client and end-points?

**Reply by infinitydon** [reply]
Applications am using:

Open5gs AMF
UERANSIM for the gNB simulation

**Reply by ghost** [reply]
Confirmed to be working in our setup. Just needed  the following in (sctp client and end-point host) :

```
echo Y > /sys/module/sctp/parameters/no_checksums 
``` 

Also, please note that "tcpdump -i any host 192.168.1.125" will show only the incoming request as the outgoing request in full nat mode will have source/dest both changed.

**Comment by TrekkieCoder** [comment]
Hi @infinitydon 

Thanks for your time today. I think there was some confusion related to nat-mode implementation in loxilb. So,  we would like to clarify mode details. As per your scenario :

```mermaid

sequenceDiagram
    participant C1
    participant loxilb
    participant EP1
    C1->>loxilb: 192.168.56.X<->192.168.56.101
    loxilb->>EP1: 192.168.70.1<->192.168.70.3
```

1. Full NAT mode

```
loxicmd create lb 88.88.88.1 --sctp=38412:38412 --endpoints=192.168.70.3:1 --mode=fullnat
```
In this case, traffic with a source IP of 88.88.88.1 (after NAT) will be sent towards end-point 192.168.70.3 and this end-point is expected to have a route of 88.88.88.1 back to loxilb. We have this mode to be able to use BGP to influence route back to loxilb in case of HA/Cluster deployment of loxilb on a per-rule basis.

2. One ARM mode

```
loxicmd create lb 88.88.88.1 --sctp=38412:38412 --endpoints=192.168.70.3:1 --mode=onearm
```
In this case, traffic with a source IP of 192.168.70.1 (interface IP) will be sent  towards end-point 192.168.70.3. Since end-point will be in the same subnet, there is no need for any route configuration at end-point.  Also, for this mode it is not necessary to have Client C1 to be in the same subnet. If Client C1 is to be in the same subnet, only additional requirement is to disable rp_filter in loxilb (at least in the interfaces used for LB operation) and make the additional topology changes. 

```
for i in /proc/sys/net/ipv4/conf/*/rp_filter; do echo 0 > "$i"; done
```

I hope it helps !!

**Reply by infinitydon** [reply]
Thanks for the time, it is clearer now. I have shared my learning in my second LoxiLB blog-post:

https://futuredon.medium.com/5g-sctp-loadbalancer-using-loxilb-b525198a9103

**Reply by infinitydon** [reply]
By the way, what is required to enable the bgp mode, I want to advertise the VIP to a BGP router (FRR)

**Reply by ghost** [reply]
You can find information on how to run bgp mode [here](https://github.com/loxilb-io/loxilbdocs/blob/main/docs/integrate_bgp_eng.md) Section 4 onwards. In summary :
1. Need gobgp installed. If using loxilb docker, it is already pre-installed. (expected location : /usr/sbin/gobgp)
2. Need to run loxilb with a "-b" option. Pass it as argument to docker or  run it directly as command-line args if not using docker.
3. Update bgp peering [information](https://github.com/loxilb-io/loxilbdocs/blob/main/docs/integrate_bgp_eng.md#42-gobgp_loxilbyaml-file-setup) and restart loxilb
4. Finally to advertise the VIP,  need to add the rule also with --bgp option. For example - 
```
loxicmd create lb 20.20.20.1 --tcp=2020:8080 --endpoints=31.31.31.1:1 --bgp 
```
It is also possible to use gobgp utils to advertise VIP/route without using "loxicmd --bgb" . For example -
```
gobgp global rib add 20.20.20.1/32 -a ipv4
```
